backupccl: allow online restore to assume exclusive end keys

stevendanna opened the pull request:
Currently, we only support non-revision-history restores for online restore. As a result, we can assume file span end keys are exclusive. We were implicitly assuming this and throwing away files that didn't overlap with the restore span entry.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114464)
<!-- Reviewable:end -->

stevendanna committed:
backupccl: allow online restore to assume exclusive end keys

Currently, we only support non-revision-history restores for online
restore. As a result, we can assume file span end keys are
exclusive. We were implicitly assuming this and throwing away files
that didn't overlap with the restore span entry.

Release note: None

stevendanna commented:
eek, bad rebase.

msbutler commented on pkg/ccl/backupccl/restore_span_covering.go:
```diff
@@ -635,8 +637,23 @@ func getNewIntersectingFilesByLayer(
+
+// exclusiveEndKeyComparator assumes that file spans have exclusive
+// end keys.
+type exclusiveEndKeyComparator struct{}
```
don't we need to pass this to the input of genSpans used by online restore?

dt commented:
bors r=dt
